src/psyclone/transformations.py

```diff
@@ -2630,6 +2676,207 @@ def validate(self, kern):
+
+        '''
+        self._validate(node_list, default_present)
+
+        # Keep a record of this transformation
+        from psyclone.undoredo import Memento
+        keep = Memento(node_list[:], self)
+
+        parent = node_list[0].parent
+        schedule = node_list[0].root
+
+        # Create the directive and insert it. Take a copy of the list
+        # as it may just be a reference to the parent.children list
+        # that we are about to modify.
+        from psyclone.psyGen import ACCKernelsDirective
+        directive = ACCKernelsDirective(parent=parent,
+                                        children=node_list[:],
+                                        default_present=default_present)
+        start_index = parent.children.index(node_list[0])
+
+        for child in directive.children:
+            parent.children.remove(child)
+            child.parent = directive
+
+        parent.children.insert(start_index, directive)
+
+        # Return the now modified kernel
+        return schedule, keep
+
+    def _validate(self, node_list, default_present):
+        '''
+        Check that we can safely enclose the supplied list of nodes within
+        OpenACC kernels ... end kernels directives.
+        '''
+        from psyclone.nemo import NemoSchedule
+        from psyclone.psyGen import Loop
+        # Check that the API is valid
+        sched = node_list[0].root
+        if not isinstance(sched, NemoSchedule):
+            raise NotImplementedError("OpenACC kernels regions are currently "
+                                      "only supported for the nemo API")
+        super(ACCKernelsTrans, self)._validate(node_list)
+
+        # Check that we have at least one loop within the proposed region
+        found = False
+        for node in node_list:
+            loops = node.walk(node.children, Loop)
+            if loops or isinstance(node, Loop):
+                found = True
+                break
+        if not found:
+            raise TransformationError("A kernels transformation must enclose "
+                                      "at least one loop but none were found.")
+
+
+class ACCDataTrans(RegionTrans):
+    '''
+    Add an OpenACC data region around a list of nodes in the PSyIR.
+
+    For example:
+
+    >>> from psyclone.parse import parse
+    >>> from psyclone.psyGen import PSyFactory
+    >>> api = "NEMO"
+    >>> filename = "tra_adv.F90"
+    >>> ast, invokeInfo = parse(filename, api=api)
+    >>> psy = PSyFactory(api).create(invokeInfo)
+    >>>
+    >>> from psyclone.transformations import ACCDataTrans
+    >>> dtrans = ACCDataTrans()
+    >>>
+    >>> schedule = psy.invokes.get('invoke_0').schedule
+    >>> schedule.view()
+    >>> kernels = schedule.children[0].children[0].children[0:-1]
+    >>> # Enclose the kernels
+    >>> new_sched, _ = dtrans.apply(kernels)
+
+    '''
+    from psyclone import psyGen
+    valid_node_types = (psyGen.Loop, psyGen.Kern, psyGen.BuiltIn,
```

Comment:
This is specified in a previous class. Perhaps it should be static?

Reply:
Ah I see. Yes, it's a static, immutable class property that must be defined for each class.